File: 0759-set-intersection-size-at-least-two/0759-set-intersection-size-at-least-two.cpp
class Solution {
public:

    static bool compare(vector<int>&a, vector<int>&b){
        if(a[1]==b[1])return a[0]<b[0];
        else return a[1]<b[1];
    }

    int intersectionSizeTwo(vector<vector<int>>& inter) {
        int n = inter.size();

        sort(inter.begin(), inter.end(), compare);
        vector<int>res;

        res.push_back(inter[0][1]-1);
        res.push_back(inter[0][1]);

        for(int i=1;i<n;i++){
            int strt = inter[i][0];
            int end = inter[i][1];

            if(strt>res.back()){
                res.push_back(end-1);
                res.push_back(end);
            }else if(strt == res.back()){
                res.push_back(end);
            }else if(strt > res[res.size()-2]){
                res.push_back(end);
            }
        }
        return res.size();
    }
};

File: 0759-set-intersection-size-at-least-two/0759-set-intersection-size-at-least-two.java
class Solution {
    public int intersectionSizeTwo(int[][] intervals) {
    Arrays.sort(intervals, new Comparator<int[]>() {
    	public int compare(int[] a, int[] b) {
    	    return (a[1] != b[1] ? Integer.compare(a[1], b[1]) : Integer.compare(b[0], a[0]));
    	}
    });
    
    int m = 0, largest = -1, second = -1;
    
    for (int[] interval : intervals) {
        int a = interval[0], b = interval[1];
	
	    boolean is_largest_in = (a <= largest);
        boolean is_second_in = (a <= second);
	
	    if (is_largest_in && is_second_in) continue;
        
        m += (is_largest_in ? 1 : 2);
        
	    second = (is_largest_in ? largest : b - 1);
	    largest = b;
    }
    
    return m;
}
}